observer/baseNodeProvider.go

```diff
@@ -32,6 +36,82 @@ func (bop *baseNodeProvider) initNodesMaps(nodes []*data.NodeData) error {
 	return nil
 }
 
+// ReloadNodes will reload the observers or the full history observers
+func (bop *baseNodeProvider) ReloadNodes(nodesType data.NodeType) data.NodesReloadResponse {
+	bop.mutNodes.RLock()
+	oldNodes := bop.nodes
+	bop.mutNodes.RUnlock()
+
+	newConfig, err := loadMainConfig(bop.configurationFilePath)
+	if err != nil {
+		return data.NodesReloadResponse{
+			OkRequest:   true,
+			Description: "not reloaded",
+			Error:       "cannot load configuration file at " + bop.configurationFilePath,
+		}
+	}
+
+	nodes := newConfig.Observers
+	if nodesType == data.FullHistoryNode {
+		nodes = newConfig.FullHistoryNodes
+	}
+
+	newNodes := nodesSliceToShardedMap(nodes)
+	numOldShardsCount := len(oldNodes)
+	numNewShardsCount := len(newNodes)
+
+	if numOldShardsCount != numNewShardsCount {
+		return data.NodesReloadResponse{
+			OkRequest:   false,
+			Description: "not reloaded",
+			Error:       fmt.Sprintf("different number of shards. before: %d, now: %d", numOldShardsCount, numNewShardsCount),
+		}
+	}
+
+	bop.mutNodes.Lock()
+	bop.nodes = newNodes
+	bop.allNodes = initAllNodesSlice(newNodes)
+	bop.mutNodes.Unlock()
+
+	return data.NodesReloadResponse{
+		OkRequest:   true,
+		Description: prepareReloadResponseMessage(newNodes),
+		Error:       "",
+	}
+}
+
+func loadMainConfig(filepath string) (*config.Config, error) {
+	cfg := &config.Config{}
+	err := core.LoadTomlFile(cfg, filepath)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+func nodesSliceToShardedMap(nodes []*data.NodeData) map[uint32][]*data.NodeData {
+	newNodes := make(map[uint32][]*data.NodeData)
+	for _, observer := range nodes {
+		shardId := observer.ShardId
+		newNodes[shardId] = append(newNodes[shardId], observer)
+	}
+
+	return newNodes
+}
+
+func prepareReloadResponseMessage(newNodes map[uint32][]*data.NodeData) string {
```

Comment:
add unit tests for this function to ensure is working correctly.

Reply:
done